### PR TITLE
Canonicalize "__key#" to "__key." in indexes and queries

### PR DESCRIPTION
**Pull request by taburet**
Fixes: https://github.com/hazelcast/hazelcast/issues/15808

### CODE REVIEW

File: hazelcast/src/test/java/com/hazelcast/map/impl/query/QueryBasicTest.java [hunk 1]
```diff
@@ -952,6 +954,35 @@ public void testQueryPortableObjectWithIndexAndAlwaysCacheValues() {
...
+            map.put(new CompositeKey(i), i);
+        }
+
+        map.values(Predicates.equal("__key.a", 5));
+        map.values(Predicates.equal("__key#b", 5));
+        map.values(Predicates.equal("__key#c", 5));
+        map.values(Predicates.equal("__key.d", 5));
+
+        Map<String, LocalIndexStats> stats = map.getLocalMapStats().getIndexStats();
+        assertEquals(1, stats.get("__key.a").getQueryCount());
```
Comment (petrpleshachkov) [inline hazelcast/src/test/java/com/hazelcast/map/impl/query/QueryBasicTest.java]: I would also check that before queries the `stats` values are zero.

File: hazelcast/src/test/java/com/hazelcast/map/impl/query/QueryBasicTest.java [hunk 2]
```diff
@@ -971,4 +1002,34 @@ private void testQueryUsingNestedPortableObject(Config config, String name) {
         values = map.values(new SqlPredicate("child.child.timestamp > 0"));
         assertEquals(1, values.size());
     }
+
+    public static class CompositeKey implements Serializable {
+
+        private int a;
```
Comment (petrpleshachkov) [inline hazelcast/src/test/java/com/hazelcast/map/impl/query/QueryBasicTest.java]: make the fields `final`.

File: hazelcast/src/test/java/com/hazelcast/map/impl/query/QueryBasicTest.java (hunk 1 above)
Comment (taburet) [reply hazelcast/src/test/java/com/hazelcast/map/impl/query/QueryBasicTest.java]: done

File: hazelcast/src/test/java/com/hazelcast/map/impl/query/QueryBasicTest.java (hunk 2 above)
Comment (taburet) [reply hazelcast/src/test/java/com/hazelcast/map/impl/query/QueryBasicTest.java]: done

**Pull request merged**